app/src/main/java/com/google/android/sambadocumentsprovider/ShareManager.java

```diff
@@ -64,56 +66,119 @@
     // Loading mounted servers
     final Set<String> serverStringSet =
         mPref.getStringSet(SERVER_STRING_SET_KEY, Collections.<String> emptySet());
-    final Map<String, CredentialTuple> credentialMap = new HashMap<>(serverStringSet.size());
+    final Map<String, ShareTuple> shareMap = new HashMap<>(serverStringSet.size());
     for (String serverString : serverStringSet) {
       // TODO: Add decryption
-      String uri = decode(serverString, credentialMap);
+      String uri = decode(serverString, shareMap);
       if (uri != null) {
         mServerStringMap.put(uri, serverString);
       }
     }
 
     mServerStringSet = new HashSet<>(serverStringSet);
 
-    for (Map.Entry<String, CredentialTuple> server : credentialMap.entrySet()) {
-      final CredentialTuple tuple = server.getValue();
+    for (Map.Entry<String, ShareTuple> server : shareMap.entrySet()) {
+      final ShareTuple tuple = server.getValue();
+
+      if (tuple.mIsMounted) {
+        mMountedServerSet.add(server.getKey());
+      }
+
       mCredentialCache.putCredential(
           server.getKey(), tuple.mWorkgroup, tuple.mUsername, tuple.mPassword);
     }
   }
 
-  public synchronized void mountServer(
-      String uri, String workgroup, String username, String password, ShareMountChecker checker)
-      throws IOException {
-    if (mServerStringMap.containsKey(uri)) {
-      throw new IllegalStateException("Uri " + uri + " is already mounted.");
+  /**
+   * Stores the server info (e.g. uri, credentials) permanently.
+   * @param mount the share will be displayed as a separate root.
+   * @param force if the share with such a uri already exists, it will be overwritten.
+   */
+  public synchronized void addServer(
+          String uri, String workgroup, String username, String password,
+          ShareMountChecker checker, boolean mount, boolean force) throws IOException {
+    if (!force && mServerStringMap.containsKey(uri)) {
```

Comment:
I think I don't really understand what you meant by abused. We need two kind of behaviors:

1. Try to add a server and throw an error if it already exists - this is for mounting logic.
2. For auth logic, though, we need to force the update because even if the server is there the fact that AuthException has been thrown means that the credentials are out of date.

How about making the current method private and make two public ones - addServers/addOrUpdateServer that are going to use the force flag?